src/classes/frDonor.cls

```diff
@@ -64,90 +66,134 @@ public class frDonor extends frModel implements frSyncable{
-                String byAddress = 'select Id, fr_ID__c, LastName from Contact where MailingStreet = :address1' +
-                    ' and FirstName = :firstName' +
-                    ' and LastName = :lastName';
-                if(cityAndState) {
-                    byAddress += ' and MailingCity = :city' +
-                    ' and MailingState = :state';
+    
+            String mailingStreetField = isPersonAccount ? 'PersonMailingStreet' : 'MailingStreet';
+            String mailingCityField = isPersonAccount ? 'PersonMailingCity' : 'MailingCity';
+            String mailingStateField = isPersonAccount ? 'PersonMailingState' : 'MailingState';
+            String mailingPostalCodeField = isPersonAccount ? 'PersonMailingPostalCode' : 'MailingPostalCode';
+    
+            if (namePresent && String.isNotBlank(address1) && (cityAndState || String.isNotBlank(postalCode))) {
+                String byAddress = 'SELECT Id, fr_ID__c, LastName FROM ' + sObjectType + 
+                    ' WHERE ' + mailingStreetField + ' = :address1 AND FirstName = :firstName AND LastName = :lastName';
+    
+                if (cityAndState) {
+                    byAddress += ' AND ' + mailingCityField + ' = :city AND ' + mailingStateField + ' = :state';
                 }
-                if(String.isNotBlank(postalCode)) {
-                    byAddress += ' and MailingPostalCode = :postalCode';
+                if (String.isNotBlank(postalCode)) {
+                    byAddress += ' AND ' + mailingPostalCodeField + ' = :postalCode';
                 }
-                byAddress += ' limit 1';
-                
-                contacts = (List<Contact>)Database.query(byAddress);
-                supporterContact = applyMatch(contacts, frId);
+                byAddress += ' LIMIT 1';
+    
+                records = Database.query(byAddress);
+                supporter = applyMatch(records, frId);
             }
         }
-        
-        if(supporterContact == null) {
-            supporterContact = new Contact(fr_Id__c = frId);
+    
+        // If no match, create a new record
+        if (supporter == null) {
+            if(isPersonAccount) {
+                supporter = new Account(fr_Id__c = frId);
+            } 
+            else {
+                supporter = new Contact(fr_Id__c = frId);
+            }
         }
-        applyMappings(supporterContact, request);
-        
-        if(String.isBlank(supporterContact.LastName)) {
-            //LastName is required.  If the mappings didn't populate it, try to populate with the institution name
-            supporterContact.LastName = String.valueOf(request.get('institutionName'));
+    
+        // Apply mappings
+        applyMappings(supporter, request);
+    
+        if (String.isBlank((String)supporter.get('LastName'))) {
+            supporter.put('LastName', String.valueOf(request.get('institutionName')));
         }
-        
+        Database.upsert(supporter, true);
```

Comment:
Can you review this saving behavior?  It looks like we upsert once here, and then have some branching logic on lines 185-188.  Additionally, it looks like the upsert that uses the fr_ID__c external ID field was removed, but I don't think it should have been removed as that is needed to not duplicate records

Reply:
Reverted behavior and uses external id field!